tester/test_find_rule.c

```diff
@@ -2,52 +2,166 @@
 #include "keybuffer.h"
 #include "key_stack.h"
 #include "trie.h"
+#include "utils.h"
 #include "sequence_transform.h"
 #include "sequence_transform_data.h"
 #include "sequence_transform_test.h"
 #include "sim_output_buffer.h"
 #include "tester_utils.h"
 #include "tester.h"
 
+#define KEY_AT(i) st_key_buffer_get_keycode(buf, (i))
+
+//////////////////////////////////////////////////////////////////
+// Converts keys from buffer to null terminated ascii string
+// KC_SPACE -> ' ' (instead of st_wordbreak_ascii)
+void key_buffer_to_space_str(const st_key_buffer_t *buf, char *str)
+{
+    const int len = buf->context_len;
+    for (int i = 0; i < len; ++i) {
+        const uint16_t key = KEY_AT(len - i - 1);
+        *str++ = key == KC_SPACE ? ' ' : st_keycode_to_char(key);
+    }
+    *str = 0;
+}
 //////////////////////////////////////////////////////////////////////
-// (partial) simulation of process_sequence_transform logic
-// to test st_find_missed_rule
-void sim_st_find_missed_rule(const uint16_t *keycodes)
+bool buf_needs_expanding(st_key_buffer_t *buf)
 {
-    missed_rule_seq[0] = 0;
-    missed_rule_transform[0] = 0;
+    for (int i = 1; i < buf->context_len; ++i) {
+        const uint16_t key = KEY_AT(i);
+        if (st_is_seq_token_keycode(key)) {
+            return true;
+        }
+    }
+    return false;
+}
+//////////////////////////////////////////////////////////////////////
+// Setup input buffer from rule transform so that
+// it's ready for the st_find_missed_rule call.
+// Fills chained_transform with an ascii string of the actual transform
+// that will be tested.
+// returns false if rule is untestable, true otherwise
+bool setup_input_from_transform(const st_test_rule_t *rule, char *chained_transform)
+{    
     st_key_buffer_t *buf = st_get_key_buffer();
-    // reset search_len_from_space by first sending a single space
-    st_key_buffer_push(buf, KC_SPACE);
-    st_find_missed_rule();
     // send input rule seq so we can get output transform to test
-    sim_st_perform(keycodes);
-    buf->context_len = 0;
-    // send the output into input buffer
-    // to simulate user typing it directly
-    char *output = sim_output_get(false);
-    for (char c = *output; c; c = *++output) {
+    sim_st_perform(rule->seq_keycodes);
+
```

Comment:
This function has a LOT more logic than I would normally want to see in a test case, but I think that is an unavoidable artifact of only having automated testing of dynamic test cases, which sometimes requires non-trivial setup.

It will be good when we add the capability to hand craft test data with a prepopulated ruleset, so that we can also have simple stable tests for finding regressions and sanity checking that our tests themselves aren't broken.

Reply:
It does. I don't like it.
But with the rule search being `tricky`, there's no other way.
Things will be a _lot_ cleaner/simpler though once the test transform strings are stored as keycode arrays instead of ascii strings.